lib/service/supervisor.go

```diff
@@ -320,7 +339,22 @@ func (s *LocalSupervisor) BroadcastEvent(event Event) {
 
 	switch event.Name {
 	case TeleportExitEvent:
-		s.signalExit()
+		// if exit event includes a context payload, it is a "graceful" exit, and
+		// we need to hold off closing the supervisor's exit context until after
+		// the graceful context has closed.  If not, it is a hard exit and we should
+		// close the context immediately.
+		if ctx, ok := event.Payload.(context.Context); ok {
```

Comment:
Could we use a payload that is less ambiguous? Say:

```suggestion
var gracefulShutdownPayload = struct{}{}
```

and then:

```suggestion
		if event.Payload == gracefulShutdownPayload {
```

Reply:
The payload being a context is not arbitrary.  It is the context passed into `process.Shutdown(...)`, and is propagated via the event to the various `Shutdown(...)` methods of teleport's sub-components.